tests/fixtures/edx/base.py

```diff
@@ -0,0 +1,58 @@
+"""Base event factory definitions"""
+
+from datetime import timezone
+
+from factory import Factory, LazyAttribute, Maybe, lazy_attribute
+from faker import Faker
+
+from ralph.models.edx.base import BaseContextModel, BaseEventModel
+
+Faker.seed(1)
+FAKE = Faker()
+
+
+class BaseContextFactory(Factory):
+    """Base context factory inherited by all context factories"""
+
+    class Meta:  # pylint: disable=missing-class-docstring
+        model = BaseContextModel
+
+    course_user_tags = FAKE.pydict(value_types=str)
+    path = f"/{FAKE.uri_path()}"
+    user_id = FAKE.random_element(elements=(FAKE.random_int(), "", None))
+    org_id = Maybe("course_user_tags", FAKE.word(), "")
+    course_id = LazyAttribute(
+        lambda self: f"course-v1:{self.org_id}+{FAKE.word()}+{FAKE.word()}"
+        if self.org_id
+        else ""
+    )
+
+
+class BaseEventFactory(Factory):
+    """Base event factory inherited by all event factories"""
+
+    class Meta:  # pylint: disable=missing-class-docstring
+        model = BaseEventModel
+
+    class Params:  # pylint: disable=missing-class-docstring
+        context_ = {}
+
+    username = FAKE.user_name() if FAKE.boolean() else ""
+    ip = FAKE.ipv4_public()
+    agent = FAKE.user_agent()
+    host = FAKE.hostname()
+    referer = FAKE.url()
+    accept_language = FAKE.locale()
+    event_source = "server"
+    time = FAKE.iso8601(tzinfo=timezone.utc)
+    page = None
+
+    @lazy_attribute
+    def context(self):
+        """Returns the context field"""
+
+        # pylint: disable=no-member, protected-access
+        field = "course_user_tags"
+        exclude = () if field in self.context_ or FAKE.boolean() else (field,)
+        BaseContextFactory._meta.exclude = exclude
```

Comment:
:warning: Think of dependency injection instead of monkeypatching your factory here!

Reply:
Thanks for the advise! Unfortunately I wasn't able to find a right way to exclude fields from the factory. But as it's not used in the test I removed this part)